nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/PublishAMQP.java

```diff
@@ -195,29 +193,85 @@ public void process(final InputStream in) throws IOException {
      * Extracts AMQP properties from the {@link FlowFile} attributes. Attributes
      * extracted from {@link FlowFile} are considered candidates for AMQP
      * properties if their names are prefixed with
-     * {@link AMQPUtils#AMQP_PROP_PREFIX} (e.g., amqp$contentType=text/xml)
+     * {@link AMQPUtils#AMQP_PROP_PREFIX} (e.g., amqp$contentType=text/xml).
+     * The header property is an exception and requires a {@link Map}, so should
+     * be passed in the following format: amqp$headers=key$value$key$value etc.
      */
     private BasicProperties extractAmqpPropertiesFromFlowFile(FlowFile flowFile) {
         Map<String, String> attributes = flowFile.getAttributes();
         AMQP.BasicProperties.Builder builder = new AMQP.BasicProperties.Builder();
         for (Entry<String, String> attributeEntry : attributes.entrySet()) {
             if (attributeEntry.getKey().startsWith(AMQPUtils.AMQP_PROP_PREFIX)) {
-                String amqpPropName = attributeEntry.getKey().split("\\" + AMQPUtils.AMQP_PROP_DELIMITER)[1];
+                String amqpPropName = attributeEntry.getKey();
                 String amqpPropValue = attributeEntry.getValue();
-                try {
-                    if (amqpPropertyNames.contains(AMQPUtils.AMQP_PROP_PREFIX + amqpPropName)) {
-                        Method m = builder.getClass().getDeclaredMethod(amqpPropName, String.class);
-                        m.invoke(builder, amqpPropValue);
-                    } else {
-                        getLogger().warn("Unrecogninsed AMQP property '" + amqpPropName + "', will ignore.");
+
+                AMQPUtils.PropertyNames propertyNames = AMQPUtils.PropertyNames.fromValue(amqpPropName);
+
+                if (propertyNames != null) {
+                    try {
+                        switch (propertyNames){
+                            case CONTENT_TYPE:
+                                builder.contentType(amqpPropValue);
+                                break;
+                            case CONTENT_ENCODING:
+                                builder.contentEncoding(amqpPropValue);
+                                break;
+                            case HEADERS:
+                                String[] s = amqpPropValue.split("\\" + AMQPUtils.AMQP_PROP_DELIMITER);
+                                Map<String, Object> headers = new HashMap<>();
+
+                                for (int i = 0; i < s.length ; i += 2){
+                                    if (i + 2 <= s.length){
+                                        headers.put(s[i], s[i + 1]);
+                                    }
+                                }
+
+                                builder.headers(headers);
+                                break;
```

Comment:
@petmit would this not potentially cause a confusing issue where a flow tries to send a header with a key of say "userId", and it is wanted in the headers but actually populates the userId property? It also means that AMQP properties essentially become "keywords" that cannot ever be used in the headers - which doesn't seem ideal